pkg/path/parser/parser_test.go

```diff
@@ -0,0 +1,245 @@
+/*
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package parser
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+)
+
+func TestParser(t *testing.T) {
+	tests := []struct {
+		input     string
+		expected  []Node
+		expectErr bool
+	}{
+		{
+			// empty returns empty
+			input:    ``,
+			expected: nil,
+		},
+		{
+			// we don't allow a leading separator
+			input:     `.spec`,
+			expected:  nil,
+			expectErr: true,
+		},
+		{
+			// we don't allow a trailing separator
+			input:     `spec.`,
+			expected:  nil,
+			expectErr: true,
+		},
+		{
+			input: `single_field`,
+			expected: []Node{
+				&Object{Value: "single_field"},
+			},
+		},
+		{
+			input: `spec.containers[name: *].securityContext`,
+			expected: []Node{
+				&Object{Value: "spec"},
+				&Object{Value: "containers"},
+				&List{KeyField: "name", Glob: true},
+				&Object{Value: "securityContext"},
+			},
+		},
+		{
+			// A quoted '*' is a fieldValue, not a glob
+			input: `spec.containers[name: "*"].securityContext`,
+			expected: []Node{
+				&Object{Value: "spec"},
+				&Object{Value: "containers"},
+				&List{KeyField: "name", KeyValue: strPtr("*"), Glob: false},
+				&Object{Value: "securityContext"},
+			},
+		},
+		{
+			input: `spec.containers[name: foo].securityContext`,
+			expected: []Node{
+				&Object{Value: "spec"},
+				&Object{Value: "containers"},
+				&List{KeyField: "name", KeyValue: strPtr("foo")},
+				&Object{Value: "securityContext"},
+			},
+		},
+		{
+			input: `spec.containers["my key": "foo bar"]`,
+			expected: []Node{
+				&Object{Value: "spec"},
+				&Object{Value: "containers"},
+				&List{KeyField: "my key", KeyValue: strPtr("foo bar")},
+			},
+		},
+		{
+			// Error: keys with whitespace must be quoted
+			input:     `spec.containers[my key: "foo bar"]`,
+			expectErr: true,
+		},
+		{
+			// Error: values with whitespace must be quoted
+			input:     `spec.containers[key: foo bar]`,
+			expectErr: true,
+		},
+		{
+			input: `spec.containers[name: ""].securityContext`,
+			expected: []Node{
+				&Object{Value: "spec"},
+				&Object{Value: "containers"},
+				&List{KeyField: "name", KeyValue: strPtr("")},
+				&Object{Value: "securityContext"},
+			},
+		},
+		{
+			// TODO: Is this useful or should we make it a parsing error?
```

Comment:
IMO it's a parsing error.

I could see someone needing to reference a nil-valued var, would we be able to have a "nil" keyword?

Part of the reason for requiring quotes post-colon is to reserve the possibility of other special keywords.

Another option could be to have an anti-quote (e.g. a backtic) to signify "this value is a JS keyword"